packages/integrations/markdoc/src/utils.ts

```diff
@@ -0,0 +1,145 @@
+import matter from 'gray-matter';
+import path from 'node:path';
+import type fsMod from 'node:fs';
+import type { ErrorPayload as ViteErrorPayload } from 'vite';
+import type { AstroInstance } from 'astro';
+import z from 'astro/zod';
+
+/**
+ * Match YAML exception handling from Astro core errors
+ * @see 'astro/src/core/errors.ts'
+ */
+export function parseFrontmatter(fileContents: string, filePath: string) {
+	try {
+		// `matter` is empty string on cache results
+		// clear cache to prevent this
+		(matter as any).clearCache();
+		return matter(fileContents);
+	} catch (e: any) {
+		if (e.name === 'YAMLException') {
+			const err: Error & ViteErrorPayload['err'] = e;
+			err.id = filePath;
+			err.loc = { file: e.id, line: e.mark.line + 1, column: e.mark.column };
+			err.message = e.reason;
+			throw err;
+		} else {
+			throw e;
+		}
+	}
+}
+
+/**
+ * Matches AstroError object with types like error codes stubbed out
+ * @see 'astro/src/core/errors/errors.ts'
+ */
+export class MarkdocError extends Error {
+	public errorCode: number;
+	public loc: ErrorLocation | undefined;
+	public title: string | undefined;
+	public hint: string | undefined;
+	public frame: string | undefined;
+
+	type = 'MarkdocError';
+
+	constructor(props: ErrorProperties, ...params: any) {
+		super(...params);
+
+		const {
+			code = 99999,
```

Comment:
What is the expectation with these error codes? Would be good to leave a comment.

Reply:
Honestly no clue 😅 We don't have a system for error codes for integrations outside of Astro core. Just used `99999` as a stand-in. I'll leave a comment for now so we can discuss best practices on discord.